code/datums/gamemode/role/time_agent.dm

```diff
@@ -0,0 +1,165 @@
+/**
+	Time agent
+
+	You've got 5 minutes of playtime to do something weird
+
+	If you succeed, the stations threat level is escalated
+
+	If you fail, something weird happens.
+
+	Are you a bad enough dude to make sure a corgi, a rubber duck, and a bucket are in the same place at the same time?
+
+**/
+
+/datum/role/time_agent
+	name = TIMEAGENT
+	id = TIMEAGENT
+	required_pref = ROLE_MINOR
+	logo_state = "time-logo"
+	var/list/objects_to_delete = list()
+
+
+/datum/role/time_agent/Greet(var/greeting,var/custom)
+	if(!greeting)
+		return
+
+	var/icon/logo = icon('icons/logos.dmi', logo_state)
+	switch(greeting)
+		if (GREET_CUSTOM)
+			to_chat(antag.current, "<img src='data:image/png;base64,[icon2base64(logo)]' style='position: relative; top: 10;'/> <span class='danger'>[custom]</span>")
+		else
+			to_chat(antag.current, "<img src='data:image/png;base64,[icon2base64(logo)]' style='position: relative; top: 10;'/> <span class='danger'>You are a Time Agent.<br>Specifically you are a scientist by the name of John Beckett, having discovered a method to travel through time, and becoming lost to it. <br>\
+			Now, you are forced to take responsibility for maintaining the time stream by the mysterious 'Time Agency'.<br>\
+			You only have a limited amount of time before this timeline is deemed lost, in which case you will be forcibly extracted and the mission considered a failure.<br>\
+			Locate certain time-sensitive objects scattered around the station, so as to locate the time anomaly and use it for extraction.<br>\
+			This may not be the first time you visit this timeline, and it may not be the last.</span>")
+
+	to_chat(antag.current, "<span class='danger'>Remember that the items you are provided with are largely non-expendable. Try not to lose them, especially the jump charge, as it is your ticket home.</span>")
+
+
+/datum/role/time_agent/ForgeObjectives()
+	AppendObjective(/datum/objective/target/locate/random)
+	if(prob(30))
+		AppendObjective(/datum/objective/target/assassinate)
+	AppendObjective(/datum/objective/freeform/aid)
+
+
+/datum/role/time_agent/process()
+	var/list/datum/objective/jecties = objectives.GetObjectives()
+	if(!jecties.len || locate(/datum/objective/time_agent_extract) in objectives.GetObjectives())
+		return //Not set up yet
+	var/finished = TRUE
+	for(var/datum/objective/O in objectives.GetObjectives())
+		if(!O.IsFulfilled())
+			finished = FALSE
+			break
+	if(finished)
```

Comment:
All of this logic would fit well in stage() which may have not existed at the time. That way you can conveniently stage the agent to leave if necessary, and have a fail stage too